helix-core/src/main/java/org/apache/helix/spectator/RoutingTableProvider.java

```diff
@@ -529,68 +791,95 @@ private void updateCurrentStatesListeners(List<LiveInstance> liveInstances,
+        RoutingTable newRoutingTable = new CustomizedViewRoutingTable(propertyType, type);
+        _routingTableRefMap.get(key).set(newRoutingTable);
+      } else {
+        RoutingTable newRoutingTable = new RoutingTable(propertyType);
+        _routingTableRefMap.get(key).set(newRoutingTable);
+      }
+    }
   }
 
-  protected void refresh(List<ExternalView> externalViewList, NotificationContext changeContext) {
+  protected void refresh(List<ExternalView> externalViewList,
+      List<CustomizedView> customizedViewList, NotificationContext changeContext,
+      String referenceKey) {
     HelixDataAccessor accessor = changeContext.getManager().getHelixDataAccessor();
     PropertyKey.Builder keyBuilder = accessor.keyBuilder();
 
     List<InstanceConfig> configList = accessor.getChildValues(keyBuilder.instanceConfigs());
     List<LiveInstance> liveInstances = accessor.getChildValues(keyBuilder.liveInstances());
-    refresh(externalViewList, configList, liveInstances);
+    refresh(externalViewList, customizedViewList, configList, liveInstances, referenceKey);
   }
 
   protected void refresh(Collection<ExternalView> externalViews,
-      Collection<InstanceConfig> instanceConfigs, Collection<LiveInstance> liveInstances) {
+      Collection<CustomizedView> customizedViews, Collection<InstanceConfig> instanceConfigs,
+      Collection<LiveInstance> liveInstances, String referenceKey) {
     long startTime = System.currentTimeMillis();
-    RoutingTable newRoutingTable = new RoutingTable(externalViews, instanceConfigs, liveInstances);
-    resetRoutingTableAndNotify(startTime, newRoutingTable);
+    PropertyType propertyType = _routingTableRefMap.get(referenceKey).get().getPropertyType();
+    if (propertyType == PropertyType.CUSTOMIZEDVIEW) {
+      String customizedStateType = _routingTableRefMap.get(referenceKey).get().getCustomizedStateType();
+      RoutingTable newRoutingTable = new CustomizedViewRoutingTable(customizedViews,
+          instanceConfigs, liveInstances, propertyType, customizedStateType);
+      resetRoutingTableAndNotify(startTime, newRoutingTable, referenceKey);
+    } else {
+      RoutingTable newRoutingTable = new RoutingTable(externalViews, instanceConfigs,
+          liveInstances, propertyType);
+      resetRoutingTableAndNotify(startTime, newRoutingTable, referenceKey);
+    }
   }
 
   protected void refresh(Map<String, Map<String, Map<String, CurrentState>>> currentStateMap,
-      Collection<InstanceConfig> instanceConfigs, Collection<LiveInstance> liveInstances) {
+      Collection<InstanceConfig> instanceConfigs, Collection<LiveInstance> liveInstances,
+      String referenceKey) {
     long startTime = System.currentTimeMillis();
     RoutingTable newRoutingTable =
         new RoutingTable(currentStateMap, instanceConfigs, liveInstances);
-    resetRoutingTableAndNotify(startTime, newRoutingTable);
+    resetRoutingTableAndNotify(startTime, newRoutingTable, referenceKey);
```

Comment:
Isn't is fixed for current state? Why need to pass it in?

Reply:
we need referenceKey to not what routingTableRef should be reseted. However, I agree with you we can infer referenceKey. But since we have it already in routerUpdater, I think it does not make huge difference. Also it is consistent with other refresh calls as well.